File: Assets/Scripts/Lua/LuaCallCS.lua
-- Elapsed time at first time is more than next times.
-- Elapsed time is calculated by repeating 100000 times. 

LuaCallCS = Class("LuaCallCS", XLuaTestCase) 

-- Elapsed milliseconds about 0.0019
-- 0B GC
function LuaCallCS:TestNumber()
    self.csharpTestCase:TestNumber(10);
end

-- Elapsed milliseconds about 0.0061
-- 0B GC
-- LuaCallCS.obj1 = CS.UnityEngine.GameObject() -- No issue.
LuaCallCS.obj1 = {} -- Issue: lua table to csharp object will cause a lot of memory ??
function LuaCallCS:TestObject()
    self.csharpTestCase:TestObject(self.obj1)
end

-- Elapsed milliseconds about 0.0019
-- 0B GC
function LuaCallCS:TestOutRef()
    local a, b
    a, b = self.csharpTestCase:TestOutRef(a)
    -- Log.Debug("a "..tostring(a).."b "..tostring(b))
end

-- Elapsed milliseconds about 0.0019
-- 0B GC
LuaCallCS.obj2 = {}
function LuaCallCS:TestOverload()
    self.csharpTestCase:OverloadA(1)
    -- self.csharpTestCase:OverloadB(self.obj2)
end

File: Assets/Scripts/Tests/LuaCallCS.cs
using XLua;
using UnityEngine;

namespace XLuaExamples
{
    [LuaCallCSharp]
    public class LuaCallCS : XLuaTestCaseForLua<LuaCallCS>
    { 
        protected override string LuaFile { get { return "LuaCallCS"; } }
        
        // 0B GC
        public void TestNumber(int num)
        {
        }

        // If obj is LuaTable: 32B GC
        public void TestObject(object obj)
        {
            // If obj is LuaTable 
            using (LuaTable table = obj as LuaTable)
            {
            }
        }

        // 0B GC
        public void TestOutRef(ref int a, out int b)
        {
            a = 1;
            b = 2;
        }

        public void TestOverload() {}
        public void OverloadA(int a)
        {
            // Debug.Log(string.Format("int a = {0}", a));
        }
        public void OverloadB(LuaTable a)
        {
            // Debug.Log(string.Format("LuaTable a = {0}", a.ToString()));
        }
    }
}